zebrad/src/components/inbound/downloads.rs

```diff
@@ -0,0 +1,206 @@
+where
+    ZN: Service<zn::Request, Response = zn::Response, Error = BoxError> + Send + 'static,
+    ZN::Future: Send,
+    ZV: Service<Arc<Block>, Response = block::Hash, Error = BoxError> + Send + Clone + 'static,
+    ZV::Future: Send,
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + 'static,
+    ZS::Future: Send,
+{
+    type Item = Result<block::Hash, BoxError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        // This would be cleaner with poll_map #63514, but that's nightly only.
+        if let Some(join_result) = ready!(this.pending.poll_next(cx)) {
+            match join_result.expect("block download and verify tasks must not panic") {
+                Ok(hash) => {
+                    this.cancel_handles.remove(&hash);
+                    Poll::Ready(Some(Ok(hash)))
+                }
+                Err((e, hash)) => {
+                    this.cancel_handles.remove(&hash);
+                    Poll::Ready(Some(Err(e)))
+                }
+            }
+        } else {
+            Poll::Ready(None)
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.pending.size_hint()
+    }
+}
+
+impl<ZN, ZV, ZS> Downloads<ZN, ZV, ZS>
+where
+    ZN: Service<zn::Request, Response = zn::Response, Error = BoxError> + Send + 'static,
+    ZN::Future: Send,
+    ZV: Service<Arc<Block>, Response = block::Hash, Error = BoxError> + Send + Clone + 'static,
+    ZV::Future: Send,
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + 'static,
+    ZS::Future: Send,
+{
+    /// Initialize a new download stream with the provided `network` and
+    /// `verifier` services.
+    ///
+    /// The [`Downloads`] stream is agnostic to the network policy, so retry and
+    /// timeout limits should be applied to the `network` service passed into
+    /// this constructor.
+    pub fn new(network: ZN, verifier: ZV, state: ZS) -> Self {
+        Self {
+            network,
+            verifier,
+            state,
+            pending: FuturesUnordered::new(),
+            cancel_handles: HashMap::new(),
+        }
+    }
+
+    /// Queue a block for download and verification.
+    ///
+    /// This method waits for the network to become ready, and returns an error
+    /// only if the network service fails. It returns immediately after queuing
+    /// the request.
+    #[instrument(skip(self))]
+    pub async fn download_and_verify(&mut self, hash: block::Hash) -> Result<bool, Report> {
+        if self.cancel_handles.contains_key(&hash) {
+            return Ok(false);
+        }
+
+        // Check if the block is already in the state.
+        let block_state_req = self
+            .state
+            .ready_and()
+            .await
+            .map_err(|e| eyre!(e))?
+            .call(zs::Request::Block(zs::HashOrHeight::from(hash)))
```

Comment:
This line is the cause of a build problem when we do `boxed()` in the Inbound `AdvertiseBlock` function at https://github.com/ZcashFoundation/zebra/blob/4425aeee0d35ddfe338b1e1d3ad7e9187ff385ab/zebrad/src/components/inbound.rs#L218.

Complete error is:

```
error: implementation of `std::convert::From` is not general enough
   --> zebrad/src/components/inbound.rs:218:18
    |
218 |                   .boxed()
    |                    ^^^^^ implementation of `std::convert::From` is not general enough
    | 
   ::: /home/oxarbitrage/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/convert/mod.rs:386:1
    |
386 | / pub trait From<T>: Sized {
387 | |     /// Performs the conversion.
388 | |     #[cfg_attr(not(bootstrap), lang = "from")]
389 | |     #[stable(feature = "rust1", since = "1.0.0")]
390 | |     fn from(_: T) -> Self;
391 | | }
    | |_- trait `std::convert::From` defined here
    |
    = note: `std::convert::From<std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + '1)>>` would have to be implemented for the type `std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>`, for any lifetime `'1`...
    = note: ...but `std::convert::From<std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + '2)>>` is actually implemented for the type `std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + '2)>`, for some specific lifetime `'2`

```

Reply:
this error could be a compiler bug, we've had some similar issues in the past